Network/Socket/ByteString/Lazy.hs

```diff
@@ -64,7 +65,9 @@ getContents s = loop
     loop = unsafeInterleaveIO $ do
         sbs <- N.recv s defaultChunkSize
         if S.null sbs
-            then shutdown s ShutdownReceive >> return Empty
+            then do
+              shutdown s ShutdownReceive `onException` return ()
```

Comment:
I rather think that `onException` catches too many (possibly asynchronous) exceptions.  I used `catchIOError` deliberately, and even that is perhaps too broad.  We don't want to suppress timeouts, signals, which this likely does.

Reply:
Looking at more exception catching code, I think that `recvBufFrom` catches too broad a class of errors:
```haskell
        sockaddr <- peekSocketAddress ptr_sa
            `E.catch` \(E.SomeException _) -> getPeerName s
```
Contrast with `unsetIPv6Only` in `socket` (which requires `ScopedTypeVariables` as written):
```haskell
      -- The IPv6Only option is only supported on Windows Vista and later,
      -- so trying to change it might throw an error.
      E.catch (setSocketOption s IPv6Only 0) $ (\(_ :: E.IOException) -> return ())
```
Similarly, in "Unix.hsc" the error type should IMHO be narrower:
```haskell
getPeerCredential sock = E.handle (\(E.SomeException _) -> return (Nothing,Nothing,Nothing)) $ do
    (uid, gid) <- getPeerEid sock
    return (Nothing, Just uid, Just gid)
```

Reply:
It is likely worth merging this PR and opening an issue to audit exception handling more broadly in another PR.

Reply:
Yes, a separate PR would be fine.  I think 375 is ready, but perhaps we should wait for @kazu-yamamoto to confirm...